### DISCUSSION TITLE
Output over Alsa Soundserver

### DISCUSSION THREAD

**Discussion by kinderschoko** [dtcooper/raspotify#576] [Q&A] [answered]
Hi,
first of all this is a great application! I studied the wiki and everything worked nice out of the box.
I use the raspberry 2 connected via hdmi to a soundbar. The problem here is, that my soundbar have a kind of "low noise detection" or something that turns the amp off on every track skipping and then on again. This is sometimes very loud.
I wasted a lot of time to fix it, but no solution found. I created a sound with very low frequency and very low amplitude, that cant be physically played, but when i play that the amp didnt turn off any more.
Now the goal is, to play that parallel to the raspotify output.
I found out that this should be possible with the Alsa soundserver, which can mix up several sources together. I'm new with raspberry and linux so this is not easy for me. I tried the following:

**/etc/asound.conf**
`defaults.ctl.card 1`
`defaults.pcm.card 1`
`defaults.pcm.dmix.rate 48000`
`defaults.pcm.dmix.format IEC958_SUBFRAME_LE`

**Script**
**/etc/systemd/system/silentaudio.service**
`[Unit]`
`Description=Play silent Audio`
`After=raspotify.service`

`[Service]`
`ExecStart=/usr/bin/aplay -D default -t wav -r 44100 -c 2 -f IEC958_SUBFRAME_LE /dev/zero.wav`

`[Install]`
`WantedBy=multi-user.target`

I can start the script with `systemctl start silentadii`. But the Problem is, the first come, the first served. I need it both together.
How should i configure the `/etc/raspotify/conf`?

**Comment by JasonLG1979** [comment]
If your device is "dmixable" you can configure a software mixed output so multiple applications can send sound to the soundbar at the same time.

The problem being though that dmix (ALSA's software mixer) does not support `IEC958_SUBFRAME_LE`, so software mixing in bare ALSA may not be possible unless your device is capable of accepting a linear format naively.

I'm afraid without software mixing of some kind 99.9% of devices can not play audio from multiple sources simultaneously. This is not a Linux or ALSA limitation it's a hardware limitation. Unless a device has a hardware mixer (again 99.9% don't) concurrent access of the raw device is not possible.

If the device as more than 2 channels a workaround may be to split the device into 2 virtual devices with your noise going to the channels not being used by librespot.

What is the output of `aplay -L` and `aplay -l`?

**Reply by kinderschoko** [reply]
**aplay -L**
```
null
    Discard all samples (playback) or generate zero samples (capture)
default
    Playback/recording through the PulseAudio sound server
sysdefault
    Default Audio Device
lavrate
    Rate Converter Plugin Using Libav/FFmpeg Library
samplerate
    Rate Converter Plugin Using Samplerate Library
speexrate
    Rate Converter Plugin Using Speex Resampler
jack
    JACK Audio Connection Kit
oss
    Open Sound System
pulse
    PulseAudio Sound Server
upmix
    Plugin for channel upmix (4,6,8)
vdownmix
    Plugin for channel downmix (stereo) with a simple spacialization
hw:CARD=Headphones,DEV=0
    bcm2835 Headphones, bcm2835 Headphones
    Direct hardware device without any conversions
plughw:CARD=Headphones,DEV=0
    bcm2835 Headphones, bcm2835 Headphones
    Hardware device with all software conversions
sysdefault:CARD=Headphones
    bcm2835 Headphones, bcm2835 Headphones
    Default Audio Device
dmix:CARD=Headphones,DEV=0
    bcm2835 Headphones, bcm2835 Headphones
    Direct sample mixing device
usbstream:CARD=Headphones
    bcm2835 Headphones
    USB Stream Output
hw:CARD=vc4hdmi,DEV=0
    vc4-hdmi, MAI PCM i2s-hifi-0
    Direct hardware device without any conversions
plughw:CARD=vc4hdmi,DEV=0
    vc4-hdmi, MAI PCM i2s-hifi-0
    Hardware device with all software conversions
sysdefault:CARD=vc4hdmi
    vc4-hdmi, MAI PCM i2s-hifi-0
    Default Audio Device
hdmi:CARD=vc4hdmi,DEV=0
    vc4-hdmi, MAI PCM i2s-hifi-0
    HDMI Audio Output
dmix:CARD=vc4hdmi,DEV=0
    vc4-hdmi, MAI PCM i2s-hifi-0
    Direct sample mixing device
usbstream:CARD=vc4hdmi
    vc4-hdmi
    USB Stream Output
```
**aplay -l**
```
**** List of PLAYBACK Hardware Devices ****
card 0: Headphones [bcm2835 Headphones], device 0: bcm2835 Headphones [bcm2835 Headphones]
  Subdevices: 8/8
  Subdevice #0: subdevice #0
  Subdevice #1: subdevice #1
  Subdevice #2: subdevice #2
  Subdevice #3: subdevice #3
  Subdevice #4: subdevice #4
  Subdevice #5: subdevice #5
  Subdevice #6: subdevice #6
  Subdevice #7: subdevice #7
card 1: vc4hdmi [vc4-hdmi], device 0: MAI PCM i2s-hifi-0 [MAI PCM i2s-hifi-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Reply by JasonLG1979** [reply]
The `vc4hdmi` driver doesn't support software mixing. I found that out while testing the [`asound.conf` Wizard script I wrote](https://github.com/dtcooper/raspotify/blob/master/asound_conf_setup.py) (which I am in the process of rewriting in Rust. The python script is kinda a prototype/proof of concept but it still works) 

What you want to do is comment out `dtoverlay=vc4-kms-v3d` in your Pi's `/boot/config.txt` file. Assuming your using your Pi in headless mode (no desktop environment) you don't need fancy graphics drivers anyway. What that does is use the old driver which exposes the HDMI as `hw:CARD=b1,DEV=0` which does support software mixing.

You can then follow the direction in [Basic Setup Guide](https://github.com/dtcooper/raspotify/wiki/Basic-Setup-Guide#the-easy-way). Basically run the script follow the prompts and you're set.

**Reply by JasonLG1979** [reply]
The setting I would recommend for that card are:
PCM: `hw:CARD=b1,DEV=0`
Format: `S16_LE`
Sampling Rate: `44100`
Channels: `2`
Sample Rate Converter: `speexrate_medium`

**Reply by kinderschoko** [reply]
the first steps are done :) everything is installed and with the previous settings raspotify is playing sound as before.  
But now i stuck on this line:
`> PCM: hw:CARD=b1,DEV=0`
du you have an exampe for **/etc/asound.conf** where this is implemented?

**Reply by JasonLG1979** [reply]
Run the script in the wiki and follow the prompts. The script writes a asound.conf for you. Once you do that it becomes your default output and will be configured correctly and you can just run librespot and your noise service without a device arg.

**Reply by JasonLG1979** [reply]
Change the format to S16_LE in your noise service.

**Reply by kinderschoko** [reply]
@JasonLG1979 you are a genius! Thx! 🥇 
Its working fine. No more annoying clicks from my soundbar. Perfect. 

My "silence" soundfile has actually a lengths of 60s. Any idea how to loop the soundfile? I tried the following but it plays still just 60s once. Then i have to restart the script again to avoid the clicks. Anyway its not an elegant way too. But i didnt find an arg vor `aplay` to loop a soundfile.
```
[Unit]
Description=Remove that clicking noice between songs that Andy hated
After=raspotify.service

[Service]
while true; do
        ExecStart=/usr/bin/aplay -t wav -r 44100 -c 2 -f S16_LE /home/pi/media/silence.wav
done

[Install]
WantedBy=multi-user.target
```

**Reply by JasonLG1979** [reply]
You can stream straight from dev/null you don't even need a file. I don't remember how off the top of my head but I'll find out when I get home from work here in a couple hours.

**Reply by JasonLG1979** [reply]
I'm sorry that's /dev/zero not null. /dev/zero give you an infinite amount of zeros which equals silence for an audio file.

**Reply by JasonLG1979** [reply]
`aplay -r 44100 -f S16_LE /dev/zero`

Should do it.

**Reply by kinderschoko** [reply]
You made my day! Its working now 💯

**Reply by JasonLG1979** [reply]
It's actually not an uncommon problem. I will document this in the wiki for anyone else that may have this issue.

And you're for sure welcome. I'm glad I could help and that the issue was not to difficult.

**Reply by JasonLG1979** [reply]
The problem comes from poorly designed hardware that does not mute the outputs during power on and off.

**Comment by kinderschoko** [comment]
Damn i run into a new problem using the sollution above. And its even more terrifying.
The volume setting of my soundbar is on 5 (max is 50), the spotify app and the raspberry are on 100%.
When i stop my service with `systemctl stop silentaudio` i have the loud clicks on every start and track skip 
for some milliseconds much louder then the setting of 5.
When i start my service with `systemctl start silentaudio` i dont have the noisy clicks on every start and track skip. Thats cool!
But when i pause a playlist a have something like a bomb that explodes when i click on play again. 
The volume goes up and stays on this high level while the soundbar shows still 5. When i set the volume + or - 1 it goes to normal level again.
How is this possible, when all sources are on 100%?

**Reply by kinderschoko** [reply]
I figure out that this couses the problem:
`ExecStart=/usr/bin/aplay -r 44100 -f S16_LE /dev/zero`

This works on all conditions that i tried, but just for the length of my wave-file:
`ExecStart=/usr/bin/aplay -t wav -r 44100 -c 2 -f S16_LE /home/pi/MeinOrdner/silence.wav`

So i just need to loop my wave-file. But how?

**Reply by JasonLG1979** [reply]
You certainly seem to have a misbehaving device,lol!!! Completely disable volume control in librespot by changing:

`#LIBRESPOT_INITIAL_VOLUME="50"` to `LIBRESPOT_INITIAL_VOLUME="100"` 

and

`#LIBRESPOT_VOLUME_CTRL="log"` to `LIBRESPOT_VOLUME_CTRL="fixed"`

And see what happens.

**Reply by JasonLG1979** [reply]
Doing that makes volume a No-op in librespot.

**Reply by JasonLG1979** [reply]
Otherwise you can just make your service call a shell script.

Something like:

```bash
#!/bin/bash
while [ 1 ] ; do
 aplay /home/pi/MeinOrdner/silence.wav
done
```

Should do it.

**Reply by kinderschoko** [reply]
Yes you are right. Never buy soundbars from Teufel again :D 
I tried that. Unfortunately it has no effect on this problem. 
I noticed an interesting correlation: 
When i click on pause and on play again within under 5s, everything is nice. Pause over 5s an then play: The neighbor appears to knocking on my door.
The option with playing a .wav is not as good as i expected: The soundquality of softwaremixing is poor on my raspberry :(
But there are no other problems.

Is there a other possibility? 
Get the play and pause state from libre and play a silent wave on pause and play this `aplay -r 44100 -f S16_LE /dev/zero` on regular playback?
I know this is really messy.

**Reply by JasonLG1979** [reply]
Sounds like overzealous power management in the sound bar itself maybe? Like it's turning itself off when it detects silence? There's not really anything that I know that you can do about that if there isn't a way to disable that behavior on the device itself. Sounds like you may have just bought a shitty soundbar. I'm sorry.

**Reply by JasonLG1979** [reply]
Although something you might try is using `speaker-test` to play a loop of very faint pink-noise. You'll have to inch the volume up so it reaches just above the threshold that the soundbar shuts off and see if that's acceptable. With luck the music will mask the pink-noise while it's playing.

So in your service you'd use:

`speaker-test -c1 -l0 -FS16_LE -r44100 -S1 > /dev/null 2>&1`

The `-S1` part is the volume. `S` stands for scale (0-100). Inch the Scale up until your soundbar doesn't auto shutoff.

**Reply by kinderschoko** [reply]
this is the best solution for me so far. With `-S1` the soundbar doesnt auto shutoff, but on louder volume the pink noise is clearly hearable.
Is it possible to change somewhere the datatype of `S` to float and then set on `-S0.1`?
I tried this: 
`speaker-test -c1 -l0 -FS16_LE -r44100 -S1 -t sine -X -f30000` (sinewave of 30kHz, 1%volume)
It works too, i cant hear any disturbing things. For some reason settings over 32kHz are hearable again. Limitations?

**Reply by JasonLG1979** [reply]
I would be careful about pushing ultrasound to your soundbar it can damage class D amps. Especially cheap ones that likely lack the proper filtering.

**Comment by JasonLG1979** [comment, marked as answer]
30kHz is well beyond human hearing. What you're maybe hearing at 32kHz is the amp becomeing unstable because you might be hitting some harmonic of the PWM's carrier freq? Ofc at that freq everything is all harmonics 44100 sampling rate can't produce anything over 22kHz.

**Reply by kinderschoko** [reply]
my final sollution is 
`ExecStart=speaker-test -c1 -l0 -FS16_LE -r44100 -S1 -t sine -X -f10`
I tried 25Hz, this is the lowest setting where i can feel and see my sub moving. On 10Hz nothing happens. I think the amp should handle that. But the reslut is perfect for me.
Thank you @JasonLG1979!

**Reply by JasonLG1979** [reply]
Most amps unless they're DC coupled won't go down that low. I'd still check to make sure the amp isn't getting hot. You may not hear, see or feel it but it could cook your amp.

**Comment by JasonLG1979** [comment]
As far as the volume goes I'm pretty sure it's integer only.
